src/3d/qgstessellator.cpp

```diff
@@ -157,22 +157,87 @@ void QgsTessellator::addPolygon( const QgsPolygonV2 &polygon, float extrusionHei
   polyline.reserve( exterior->numPoints() );
 
   QgsVertexId::VertexType vt;
-  QgsPoint pt, ptPrev;
+  QgsPoint pt;
+
+  QVector3D pNormal( 0, 0, 0 );
+  const int pCount = exterior->numPoints();
+  for ( int i = 0; i < pCount - 1; i++ )
+  {
+    QgsPoint pt1, pt2;
+    exterior->pointAt( i, pt1, vt );
+    exterior->pointAt( i + 1, pt2, vt );
+
+    pNormal.setX( pNormal.x() + ( pt1.y() - pt2.y() ) * ( pt1.z() + pt2.z() ) );
+    pNormal.setY( pNormal.y() + ( pt1.z() - pt2.z() ) * ( pt1.x() + pt2.x() ) );
+    pNormal.setZ( pNormal.z() + ( pt1.x() - pt2.x() ) * ( pt1.y() + pt2.y() ) );
+  }
+
+  pNormal.normalize();
+
+  // Polygon is a triangle
+  if ( pCount == 4 )
+  {
+    QgsPoint pt;
+    for ( int i = 0; i < 3; i++ )
+    {
+      exterior->pointAt( i, pt, vt );
+      mData << pt.x() - mOriginX << pt.z() << - pt.y() + mOriginY;
+      if ( mAddNormals )
+        mData << pNormal.x() << pNormal.z() << - pNormal.y();
+    }
+
+    return;
+  }
+
+  if ( pNormal.length() < 0.999 || pNormal.length() > 1.001 )
+  {
+    return;
```

Comment:
This is meant to be working like `pNormal != 1`, but with a minimum tolerance. Ideally, since we have just called the `normalize` function, this should never happen. Yet, there are cases where the function fails to actually make a normal of unit length (mostly makes an almost-zero length vector), which will make later calculations to fail (divide by zero).